File: go/session/xml.go
package session

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
)

// ODict 保序字典，等价于 Python 版的 OrderedDict。
// 用于构造请求体时保持字段顺序（华为设备对部分端点字段顺序敏感）。
type ODict struct {
	keys []string
	vals []interface{}
}

// O 便捷构造：ODict 字面量。用法 O("a", 1, "b", "x")。
func O(pairs ...interface{}) *ODict {
	d := &ODict{}
	for i := 0; i+1 < len(pairs); i += 2 {
		d.Set(pairs[i].(string), pairs[i+1])
	}
	return d
}

// Set 追加或覆盖 key（保持首次插入顺序）。
func (d *ODict) Set(key string, val interface{}) {
	for i, k := range d.keys {
		if k == key {
			d.vals[i] = val
			return
		}
	}
	d.keys = append(d.keys, key)
	d.vals = append(d.vals, val)
}

// Keys 返回 key 的有序切片（副本）。
func (d *ODict) Keys() []string {
	out := make([]string, len(d.keys))
	copy(out, d.keys)
	return out
}

// Get 返回 key 的值。
func (d *ODict) Get(key string) (interface{}, bool) {
	for i, k := range d.keys {
		if k == key {
			return d.vals[i], true
		}
	}
	return nil, false
}

// Contains 报告 key 是否存在。
func (d *ODict) Contains(key string) bool {
	_, ok := d.Get(key)
	return ok
}

// Len 返回条目数。
func (d *ODict) Len() int {
	return len(d.keys)
}

// ParseXML 解析 XML 字节流为 map[string]interface{} 树，
// 等价 Python：xmltodict.parse(data, dict_constructor=dict)。
//
// 语义：
//   - 重复同名子元素 → []interface{}
//   - 叶子元素 → string
//   - 有子元素的元素 → map[string]interface{}
//   - 属性 → "@attr" 键（并入元素 dict，xmltodict 的 cdata/属性合并行为）
//   - 元素既有文本又有子元素 → "#text" 键（xmltodict 行为）
//   - 纯空白文本忽略
//
// 返回根元素名到内容的映射，如 {"response": {...}}。
func ParseXML(data []byte) (map[string]interface{}, error) {
	dec := xml.NewDecoder(bytes.NewReader(data))
	dec.Strict = false
	dec.CharsetReader = charsetReader

	var root *xmlNode
	var stack []*xmlNode
	var textBuf strings.Builder

	for {
		tok, err := dec.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("xml parse: %w", err)
		}
		switch t := tok.(type) {
		case xml.StartElement:
			textBuf.Reset()
			node := &xmlNode{
				name:  t.Name.Local,
				attrs: t.Attr,
			}
			if len(stack) == 0 {
				root = node
			} else {
				parent := stack[len(stack)-1]
				parent.children = append(parent.children, node)
			}
			stack = append(stack, node)
		case xml.CharData:
			textBuf.Write(t)
		case xml.EndElement:
			if len(stack) == 0 {
				continue
			}
			node := stack[len(stack)-1]
			node.text = textBuf.String()
			textBuf.Reset()
			stack = stack[:len(stack)-1]
		}
	}

	if root == nil {
		return map[string]interface{}{}, nil
	}

	out := map[string]interface{}{}
	if len(root.children) == 0 {
		out[root.name] = root.text
	} else {
		out[root.name] = nodeToDict(root)
	}
	return out, nil
}

// xmlNode 内部解析节点。
type xmlNode struct {
	name     string
	text     string
	children []*xmlNode
	attrs    []xml.Attr
}

// nodeToDict 将节点转为 dict 值：
//   - 有子元素 → map[string]interface{}（含 @attr 与 #text）
//   - 无子元素 → string
func nodeToDict(n *xmlNode) interface{} {
	if len(n.children) == 0 {
		if len(n.attrs) == 0 {
			return n.text
		}
		// 叶子带属性：xmltodict 返回 {"@attr":..., "#text":...}
		m := map[string]interface{}{}
		for _, a := range n.attrs {
			m["@"+a.Name.Local] = a.Value
		}
		m["#text"] = n.text
		return m
	}
	m := map[string]interface{}{}
	for _, a := range n.attrs {
		m["@"+a.Name.Local] = a.Value
	}
	if strings.TrimSpace(n.text) != "" {
		m["#text"] = n.text
	}
	for _, c := range n.children {
		val := nodeToDict(c)
		if existing, ok := m[c.name]; ok {
			switch t := existing.(type) {
			case []interface{}:
				m[c.name] = append(t, val)
			default:
				m[c.name] = []interface{}{existing, val}
			}
		} else {
			m[c.name] = val
		}
	}
	return m
}

func charsetReader(charset string, input io.Reader) (io.Reader, error) {
	// 设备响应基本为 UTF-8；其他编码原样透传
	return input, nil
}

// ConvertToMap 将 *ODict 转普通 map（保序信息丢失）。
func ConvertToMap(d *ODict) map[string]interface{} {
	out := make(map[string]interface{})
	for i, k := range d.keys {
		out[k] = d.vals[i]
	}
	return out
}

// MarshalRequest 将 data（*ODict/map/list/int/string 等）序列化为
// `<request>...</request>`，等价于 Python：
//
//	xmltodict.unparse({"request": data})
//
// 默认输出紧凑格式（无缩进），与 xmltodict.unparse 一致。
func MarshalRequest(data interface{}) ([]byte, error) {
	var buf bytes.Buffer
	buf.WriteString(`<?xml version="1.0" encoding="utf-8"?>`)
	if err := writeElement(&buf, "request", data, 0); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// writeElement 递归写出 XML 元素（紧凑模式）。
//
// 值类型规则（等价 xmltodict.unparse）：
//   - string → 文本
//   - *ODict / map → 子元素（map 按键排序回退保序）
//   - []*ODict / []map / []interface{} / []string / []int → 重复同名元素
//   - int / int64 / float64 → 十进制文本
//   - bool → "True" / "False"（Python str(v) 语义）
//   - nil → 空元素 `<x></x>`
func writeElement(buf *bytes.Buffer, name string, val interface{}, depth int) error {
	buf.WriteString("<" + name + ">")

	switch v := val.(type) {
	case nil:
		// 空元素，xmltodict 对 None 输出 <x></x>
	case string:
		buf.WriteString(escapeXML(v))
	case int:
		buf.WriteString(strconv.Itoa(v))
	case int64:
		buf.WriteString(strconv.FormatInt(v, 10))
	case float64:
		buf.WriteString(strconv.FormatFloat(v, 'f', -1, 64))
	case bool:
		if v {
			buf.WriteString("True")
		} else {
			buf.WriteString("False")
		}
	case *ODict:
		for i, k := range v.keys {
			if err := writeElement(buf, k, v.vals[i], depth+1); err != nil {
				return err
			}
		}
	case map[string]interface{}:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			if err := writeElement(buf, k, v[k], depth+1); err != nil {
				return err
			}
		}
	case map[string]string:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			if err := writeElement(buf, k, v[k], depth+1); err != nil {
				return err
			}
		}
	case []interface{}:
		for _, item := range v {
			if err := writeElement(buf, name, item, depth+1); err != nil {
				return err
			}
		}
	case []*ODict:
		for _, item := range v {
			if err := writeElement(buf, name, item, depth+1); err != nil {
				return err
			}
		}
	case []map[string]interface{}:
		for _, item := range v {
			if err := writeElement(buf, name, item, depth+1); err != nil {
				return err
			}
		}
	case []string:
		for _, item := range v {
			if err := writeElement(buf, name, item, depth+1); err != nil {
				return err
			}
		}
	case []int:
		for _, item := range v {
			if err := writeElement(buf, name, item, depth+1); err != nil {
				return err
			}
		}
	default:
		return fmt.Errorf("unsupported xml value type %T for %s", v, name)
	}

	buf.WriteString("</" + name + ">")
	return nil
}

// escapeXML 对 XML 文本做转义（等价 xmltodict 的转义，
// 注意 xmltodict 不转义单引号）。
func escapeXML(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s {
		switch r {
		case '&':
			b.WriteString("&amp;")
		case '<':
			b.WriteString("&lt;")
		case '>':
			b.WriteString("&gt;")
		case '"':
			b.WriteString("&quot;")
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

File: src/xml.rs
//! Lightweight XML builder/parser used to talk to the Huawei CPE.
//!
//! The Python library uses `xmltodict` to turn a Python dict into XML for
//! request bodies (`<request>...</request>`) and to parse XML responses back
//! into dicts. This module re-implements exactly that behaviour in Rust:
//!
//! * [`XmlValue`] is a node in the tree (a scalar string, a nested map, or a
//!   list of nodes).
//! * [`XmlMap`] is a `BTreeMap<String, XmlValue>` — ordered, which matters
//!   because some devices are sensitive to the order of the fields
//!   (see `sms/sms-list` in the Python library).
//! * [`to_xml`] serializes a map into the `<request>` XML document.
//! * [`from_xml`] parses an XML response into an `XmlMap`, collapsing
//!   single-child elements and producing lists for repeated elements, exactly
//!   like `xmltodict`.

use std::collections::BTreeMap;
use std::fmt::Write as _;

use crate::errors::{Error, Result};

/// An ordered XML attribute map (field name → value).
pub type XmlMap = BTreeMap<String, XmlValue>;

/// A node in an XML tree, mirroring what `xmltodict` produces.
#[derive(Debug, Clone, PartialEq)]
pub enum XmlValue {
    /// A leaf: text content (always a string, like Python).
    Text(String),
    /// A nested element with children.
    Map(XmlMap),
    /// A repeated element (Python `xmltodict` produces a list).
    List(Vec<XmlValue>),
    /// Explicit empty element (`<tag/>`).
    Empty,
}

impl XmlValue {
    /// Coerce to a string, treating `None`/`Empty` as empty string, like the
    /// Python code `str(data.get(key, ""))`.
    pub fn as_str(&self) -> String {
        match self {
            XmlValue::Text(s) => s.clone(),
            XmlValue::Empty => String::new(),
            XmlValue::Map(_) | XmlValue::List(_) => String::new(),
        }
    }

    /// Try to parse as an integer (default 0 on failure), like the Python
    /// `int(data.get(key, 0))` patterns.
    pub fn as_int(&self) -> i64 {
        self.as_str().trim().parse().unwrap_or(0)
    }

    /// True if this value represents the string `"1"` (used for boolean flags).
    pub fn is_truthy(&self) -> bool {
        self.as_str() == "1"
    }
}

/// Escape a string for use inside XML text content.
pub fn escape_text(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        match c {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&apos;"),
            _ => out.push(c),
        }
    }
    out
}

/// Serialize a single `XmlValue` under `tag` into `writer`.
///
/// Lists are emitted as repeated `<tag>...</tag>` elements (matching
/// `xmltodict`'s handling of a Python list).
fn write_value(tag: &str, value: &XmlValue, depth: usize, out: &mut String) {
    let indent = "  ".repeat(depth);
    match value {
        XmlValue::Empty => {
            let _ = writeln!(out, "{indent}<{tag}/>");
        }
        XmlValue::Text(text) => {
            let _ = writeln!(out, "{indent}<{tag}>{}</{tag}>", escape_text(text));
        }
        XmlValue::Map(map) => {
            let _ = writeln!(out, "{indent}<{tag}>");
            for (child_tag, child) in map {
                write_value(child_tag, child, depth + 1, out);
            }
            let _ = writeln!(out, "{indent}</{tag}>");
        }
        XmlValue::List(items) => {
            for item in items {
                write_value(tag, item, depth, out);
            }
        }
    }
}

/// Serialize an `XmlMap` into a `<request>` XML document (UTF-8).
///
/// The result is later encoded to CESU-8 by the session layer (mirroring
/// `cesu8_encode(xmltodict.unparse(...))` in Python).
pub fn to_xml(data: &XmlMap) -> String {
    let mut out = String::new();
    out.push_str("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<request>\n");
    for (tag, value) in data {
        write_value(tag, value, 1, &mut out);
    }
    out.push_str("</request>");
    out
}

/// Parse a UTF-8 XML document into an `XmlMap`, mirroring `xmltodict`:
///
/// * A leaf element `<Name>x</Name>` becomes `{"Name": Text("x")}`.
/// * An element with children `<A><B>1</B></A>` becomes
///   `{"A": Map({"B": Text("1")})}`.
/// * Repeated sibling elements become a `List`.
/// * An element with attributes keeps them under `@attr` keys and its text
///   under `#text` (rare in Huawei responses).
///
/// The returned map has the document root element as its single key, e.g.
/// `{"response": {...}}` — exactly like `xmltodict.parse`.
pub fn from_xml(input: &[u8]) -> Result<XmlMap> {
    let text = std::str::from_utf8(input)
        .map_err(|e| Error::Protocol(format!("invalid UTF-8 in XML response: {e}")))?;
    let mut reader = quick_xml::Reader::from_str(text);
    reader.config_mut().trim_text(false);

    // Stack of (tag, children) for the elements we are currently inside.
    let mut stack: Vec<(String, XmlMap)> = Vec::new();
    let mut root: Option<XmlMap> = None;

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Ok(quick_xml::events::Event::Start(e)) => {
                let name = String::from_utf8_lossy(e.name().as_ref()).to_string();
                stack.push((name, XmlMap::new()));
            }
            Ok(quick_xml::events::Event::Empty(e)) => {
                let name = String::from_utf8_lossy(e.name().as_ref()).to_string();
                let mut children = XmlMap::new();
                for attr in e.attributes().flatten() {
                    let key = format!("@{}", String::from_utf8_lossy(attr.key.as_ref()));
                    children.insert(
                        key,
                        XmlValue::Text(
                            attr.unescape_value()
                                .map_err(|err| {
                                    Error::Protocol(format!("XML attribute unescape error: {err}"))
                                })?
                                .to_string(),
                        ),
                    );
                }
                if children.is_empty() {
                    if let Some((_, parent)) = stack.last_mut() {
                        insert(parent, &name, XmlValue::Empty);
                    } else {
                        let mut m = XmlMap::new();
                        m.insert(name, XmlValue::Empty);
                        root = Some(m);
                    }
                } else {
                    let child = XmlValue::Map(children);
                    if let Some((_, parent)) = stack.last_mut() {
                        insert(parent, &name, child);
                    } else {
                        let mut m = XmlMap::new();
                        m.insert(name, child);
                        root = Some(m);
                    }
                }
            }
            Ok(quick_xml::events::Event::Text(e)) => {
                let t = e
                    .unescape()
                    .map_err(|e| Error::Protocol(format!("XML text unescape error: {e}")))?;
                let t = t.trim().to_string();
                if !t.is_empty() {
                    if let Some((_, children)) = stack.last_mut() {
                        insert(children, "#text", XmlValue::Text(t));
                    }
                }
            }
            Ok(quick_xml::events::Event::End(_)) => {
                if let Some((tag, children)) = stack.pop() {
                    let child = fold_element(children);
                    if let Some((_, parent_children)) = stack.last_mut() {
                        insert(parent_children, &tag, child);
                    } else {
                        // We popped the root element.
                        let mut m = XmlMap::new();
                        m.insert(tag, child);
                        root = Some(m);
                    }
                }
            }
            Ok(quick_xml::events::Event::Eof) => break,
            Err(e) => {
                return Err(Error::Xml(e));
            }
            _ => {}
        }
        buf.clear();
    }

    Ok(root.unwrap_or_default())
}

/// Collapse a parsed element's children into an `XmlValue`:
///
/// * `{ "#text": Text("x") }` (no attribute children) → `Text("x")`
/// * empty → `Empty`
/// * anything else (has element children or attributes) → `Map`
fn fold_element(children: XmlMap) -> XmlValue {
    let has_attrs = children.keys().any(|k| k.starts_with('@'));
    if !has_attrs {
        if let Some(XmlValue::Text(t)) = children.get("#text") {
            if children.len() == 1 {
                return XmlValue::Text(t.clone());
            }
        }
        // Mixed text + child elements: keep structure, drop the stray text.
        let mut map = children;
        map.remove("#text");
        if map.is_empty() {
            return XmlValue::Empty;
        }
        return XmlValue::Map(map);
    }
    XmlValue::Map(children)
}

/// Convert a parsed JSON value into an `XmlMap`, so JSON endpoints (rare) can
/// be consumed with the same [`MapExt`] helpers as XML ones.
pub fn json_to_map(value: &serde_json::Value) -> XmlValue {
    match value {
        serde_json::Value::Object(obj) => {
            let mut map = XmlMap::new();
            for (k, v) in obj {
                map.insert(k.clone(), json_to_map(v));
            }
            XmlValue::Map(map)
        }
        serde_json::Value::Array(arr) => XmlValue::List(arr.iter().map(json_to_map).collect()),
        serde_json::Value::String(s) => XmlValue::Text(s.clone()),
        serde_json::Value::Number(n) => XmlValue::Text(n.to_string()),
        serde_json::Value::Bool(b) => XmlValue::Text(if *b { "1".into() } else { "0".into() }),
        serde_json::Value::Null => XmlValue::Empty,
    }
}

/// Insert a value into a map, converting repeated keys into a `List`.
fn insert(map: &mut XmlMap, tag: &str, value: XmlValue) {
    match map.get_mut(tag) {
        Some(XmlValue::List(items)) => items.push(value),
        Some(existing) => {
            let prev = existing.clone();
            *existing = XmlValue::List(vec![prev, value]);
        }
        None => {
            map.insert(tag.to_string(), value);
        }
    }
}

/// Convenience accessors used across API groups.
pub trait MapExt {
    /// Get a nested map by dot-path, e.g. `get_map(&["response", "Ssids"])`.
    fn get_map(&self, path: &[&str]) -> Option<&XmlMap>;
    /// Get a string value by dot-path.
    fn get_str(&self, path: &[&str]) -> Option<String>;
}

impl MapExt for XmlMap {
    fn get_map(&self, path: &[&str]) -> Option<&XmlMap> {
        let mut cur: &XmlValue = self.get(path[0])?;
        for key in &path[1..] {
            match cur {
                XmlValue::Map(m) => cur = m.get(*key)?,
                _ => return None,
            }
        }
        match cur {
            XmlValue::Map(m) => Some(m),
            _ => None,
        }
    }

    fn get_str(&self, path: &[&str]) -> Option<String> {
        let mut cur: &XmlValue = self.get(path[0])?;
        for key in &path[1..] {
            match cur {
                XmlValue::Map(m) => cur = m.get(*key)?,
                _ => return None,
            }
        }
        Some(cur.as_str())
    }
}

/// Convert an [`XmlValue`] into a [`serde_json::Value`].
///
/// This is the bridge used to build strongly-typed response structs: API
/// methods deserialize their `XmlMap` response (or a child of it) into a
/// `#[derive(serde::Deserialize)]` struct via
/// `serde_json::from_value(map_to_json(map))`.
pub fn map_to_json(map: &XmlMap) -> serde_json::Value {
    let mut obj = serde_json::Map::with_capacity(map.len());
    for (k, v) in map {
        obj.insert(k.clone(), value_to_json(v));
    }
    serde_json::Value::Object(obj)
}

/// Convert a single [`XmlValue`] into a [`serde_json::Value`].
pub fn value_to_json(value: &XmlValue) -> serde_json::Value {
    match value {
        XmlValue::Text(s) => serde_json::Value::String(s.clone()),
        XmlValue::Empty => serde_json::Value::Null,
        XmlValue::Map(m) => map_to_json(m),
        XmlValue::List(items) => {
            serde_json::Value::Array(items.iter().map(value_to_json).collect())
        }
    }
}

/// Deserialize an `XmlMap` (or the `response` child of it) into a typed
/// struct `T`, returning a [`crate::errors::Error`] on failure.
///
/// If a `response` key is present and is a map, it is used; otherwise the
/// whole map is used. Missing fields deserialize to their serde defaults /
/// `None` for `Option` fields, so a partial or unknown schema does not panic.
pub fn deserialize_response<T>(map: &XmlMap) -> crate::errors::Result<T>
where
    T: serde::de::DeserializeOwned,
{
    let value = match map.get("response") {
        Some(XmlValue::Map(m)) => map_to_json(m),
        _ => map_to_json(map),
    };
    serde_json::from_value::<T>(value).map_err(|e| {
        crate::errors::Error::Other(format!("failed to deserialize API response: {e}"))
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_to_xml_simple() {
        let mut m = XmlMap::new();
        m.insert("Username".into(), XmlValue::Text("admin".into()));
        m.insert("Password".into(), XmlValue::Text("c2VjcmV0".into()));
        let xml = to_xml(&m);
        assert!(xml.contains("<Username>admin</Username>"));
        assert!(xml.contains("<Password>c2VjcmV0</Password>"));
        assert!(xml.starts_with("<?xml"));
    }

    #[test]
    fn test_to_xml_nested_and_list() {
        let mut inner = XmlMap::new();
        inner.insert("Index".into(), XmlValue::Text("1".into()));
        let mut m = XmlMap::new();
        m.insert("Ssids".into(), XmlValue::Map(inner.clone()));
        m.insert(
            "Ssid".into(),
            XmlValue::List(vec![
                XmlValue::Map(inner.clone()),
                XmlValue::Map(inner.clone()),
            ]),
        );
        let xml = to_xml(&m);
        assert_eq!(xml.matches("<Ssid>").count(), 2);
    }

    #[test]
    fn test_escape() {
        assert_eq!(escape_text("a<b&c"), "a&lt;b&amp;c");
    }

    #[test]
    fn test_from_xml_leaf_folds_to_text() {
        let xml = br#"<?xml version="1.0" encoding="UTF-8"?><response><DeviceName>HG8121C</DeviceName></response>"#;
        let map = from_xml(xml).unwrap();
        let resp = map.get_map(&["response"]).expect("response key");
        assert_eq!(resp.get_str(&["DeviceName"]).as_deref(), Some("HG8121C"));
        assert!(matches!(
            resp.get("DeviceName"),
            Some(XmlValue::Text(t)) if t == "HG8121C"
        ));
    }

    #[test]
    fn test_from_xml_nested_map() {
        let xml =
            br#"<response><DeviceInfo><Name>A</Name><Class>B</Class></DeviceInfo></response>"#;
        let map = from_xml(xml).unwrap();
        let resp = map.get_map(&["response"]).unwrap();
        let info = resp.get_map(&["DeviceInfo"]).unwrap();
        assert_eq!(info.get_str(&["Name"]).as_deref(), Some("A"));
        assert_eq!(info.get_str(&["Class"]).as_deref(), Some("B"));
    }

    #[test]
    fn test_from_xml_repeated_becomes_list() {
        let xml = br#"<response><SmsList><Index>1</Index><Index>2</Index><Index>3</Index></SmsList></response>"#;
        let map = from_xml(xml).unwrap();
        let resp = map.get_map(&["response"]).unwrap();
        let list = resp.get("SmsList").unwrap();
        match list {
            XmlValue::Map(m) => match m.get("Index").unwrap() {
                XmlValue::List(items) => assert_eq!(items.len(), 3),
                other => panic!("expected list, got {other:?}"),
            },
            other => panic!("expected map, got {other:?}"),
        }
    }

    #[test]
    fn test_from_xml_empty_element() {
        let xml = br#"<response><Empty/></response>"#;
        let map = from_xml(xml).unwrap();
        let resp = map.get_map(&["response"]).unwrap();
        assert!(matches!(resp.get("Empty"), Some(XmlValue::Empty)));
    }

    #[test]
    fn test_roundtrip() {
        let mut inner = XmlMap::new();
        inner.insert("Index".into(), XmlValue::Text("1".into()));
        let mut m = XmlMap::new();
        m.insert("Ssids".into(), XmlValue::Map(inner.clone()));
        m.insert(
            "Ssid".into(),
            XmlValue::List(vec![
                XmlValue::Map(inner.clone()),
                XmlValue::Map(inner.clone()),
            ]),
        );
        let xml = to_xml(&m);
        let parsed = from_xml(xml.as_bytes()).unwrap();
        let req = parsed.get_map(&["request"]).unwrap();
        assert_eq!(req.get_str(&["Ssids", "Index"]).as_deref(), Some("1"));
        match req.get("Ssid").unwrap() {
            XmlValue::List(items) => assert_eq!(items.len(), 2),
            other => panic!("expected list, got {other:?}"),
        }
    }
}
